### PR TITLE
refactoring mining code

### PR DESCRIPTION
**Pull request by pragmaxim**
Reasons : 
 - `ErgoMiner` was too complex 762 LOC actor => hard to read, hard to test
 - initialization logic was mixed with core logic (candidate generation + solution validation)
 - InternalMiner  is not asking for candidates as external miners but it is being updated with them, thus it is not useful for any kind of testing of mining code
 - missing support for multiple internal miners so we would not know if changes break support of multi-miners using single node
 - use of `ask` pattern complicates things

I extracted core logic from `ErgoMiner` actor into `CandidateGenerator` actor, so now the former contains initialization logic and the latter core logic. I refactored `ErgoMiningThread` so it works the same way as external miners, so now we can be sure that if it works with `ErgoMiningThread` it should work with external miners. Support for multiple internal miners also added. I also removed some `ask` patterns which are dangerous due to working from different thread contexts.

The actors are now state-machines as there is a lot of initialization logic.

### CODE REVIEW

**Comment by kushti** [comment]
@pragmaxim please add ScalaDoc for CandidateGenerator, ErgoMiner, ErgoMiningThread and public methods there (this is a requirement for all the code touched since 2020). 

It seems some tests are failing.

**Comment by pragmaxim** [comment]
@kushti fixed, test was failing because actor system `context` is not thread-safe and it is often used when working with futures.

**Review by kushti — APPROVED** [review]
LGTM, but please check a note. 

What /mining/rewardPublicKey returns now if mining = true but wallet not initialized yet ?

File: src/main/resources/application.conf [hunk 1]
```diff
@@ -34,15 +34,21 @@ ergo {
...
     miningPubKeyHex = null
 
     # If true, a node generates blocks being offline. The only really useful case for it probably is to start a new
     # blockchain
     offlineGeneration = false
 
-    # Delay for miner after succesful block creation
-    miningDelay = 5s
+    # Interval of polling for a candidate
+    internalMinerPollingInterval = 5s
```
[the remark this answers is not in the document]
Comment (pragmaxim) [reply src/main/resources/application.conf:51]: miningDelay got replaced with internalMinerPollingInterval as the internal miner switched from pushing to polling model.

**Comment by pragmaxim** [comment]
@kushti 
>>> What /mining/rewardPublicKey returns now if mining = true but wallet not initialized yet ?

Good catch, there is an edge case when the `stashing/unstashing` would miss `ReadMinerPk` message, if the `StartMining` was not called right on the start ... I pushed a commit that should cover that.

File: src/main/scala/org/ergoplatform/http/api/MiningApiRoute.scala [hunk 2]
```diff
@@ -63,18 +64,19 @@ case class MiningApiRoute(miner: ActorRef,
   }
 
   def rewardAddressR: Route = (path("rewardAddress") & get) {
-    val addressF = (miner ? ErgoMiner.ReadMinerPk).mapTo[Option[ProveDlog]]
-      .flatMap {
-        _.fold[Future[ErgoAddress]](Future.failed(new Exception("Miner PK not initialized")))(pk => {
```
Comment (pragmaxim) [inline src/main/scala/org/ergoplatform/http/api/MiningApiRoute.scala:68]: @kushti Ok, I understand your concern, it used to throw error and now it times out ... do you want me to keep it backward compatible?

File: src/main/resources/application.conf (hunk 1 above)
[the remark this answers is not in the document]
Comment (kushti) [reply src/main/resources/application.conf:51]: I meant that *internalMinerPollingInterval* defined twice in the file

    # How frequently to ask for new block candidate
    internalMinerPollingInterval = 500ms
 
   ...

    # Interval of polling for a candidate
    internalMinerPollingInterval = 5s

**Pull request merged**